File: content/problems/2558-take-gifts-from-the-richest-pile/solutions/maximum-heap-simulation/Solution.java
class Solution {
    public long pickGifts(int[] gifts,int k){
        PriorityQueue<Integer> heap=new PriorityQueue<>(Comparator.reverseOrder());
        for(int gift:gifts)heap.offer(gift);
        while(k-->0)heap.offer((int)Math.sqrt(heap.poll()));
        long answer=0;
        for(int gift:heap)answer+=gift;
        return answer;
    }
}

File: content/problems/2558-take-gifts-from-the-richest-pile/solutions/maximum-heap-simulation/solution.cpp
class Solution {
    public:long long pickGifts(vector<int>& gifts,int k){
        priority_queue<int> heap(gifts.begin(),gifts.end());
        while(k--){
            int gift=heap.top();
            heap.pop();
            heap.push(sqrt(gift));
        }
        long long answer=0;
        while(!heap.empty()){
            answer+=heap.top();
            heap.pop();
        }
        return answer;
    }
};
